### PR TITLE
enhance std.string.indexOf(substring) with ranges

### PR DESCRIPTION
**Pull request by WalterBright**
This adds range support to overload number 3.

http://dlang.org/phobos/std_string.html#.indexOf.3

100% unit test coverage

### CODE REVIEW

File: std/string.d [hunk 1]
```diff
@@ -581,29 +581,90 @@ ptrdiff_t indexOf(Char)(const(Char)[] s, in dchar c, in size_t startIdx,
...
+                auto s2 = sbydchar.save;        // why s must be a forward range
+                foreach (c; subr.save)
+                {
+                    s2.popFront();
+                    if (s2.empty)
+                        return -1;
+                    if (cs == CaseSensitive.yes ? c != s2.front
+                                                : toLower(c) != toLower(s2.front)
+                       )
+                        goto Lnext;
```
Comment (JakobOvrum) [inline std/string.d]: Use labelled break

File: std/string.d [hunk 2]
```diff
@@ -581,29 +581,90 @@ ptrdiff_t indexOf(Char)(const(Char)[] s, in dchar c, in size_t startIdx,
...
+    Returns:
+        the index of the first occurrence of $(D sub) in $(D s). If $(D sub)
+        is not found, then $(D -1) is returned.
+        If the parameters are not valid UTF, the result will still
+        be in the range [-1 .. s.length], but will not be reliable otherwise.
   +/
-ptrdiff_t indexOf(Char1, Char2)(const(Char1)[] s, const(Char2)[] sub,
-        in CaseSensitive cs = CaseSensitive.yes) @trusted
-    if (isSomeChar!Char1 && isSomeChar!Char2)
+ptrdiff_t indexOf(Range, Char2)(Range s, const(Char2)[] sub,
```
Comment (JakobOvrum) [inline std/string.d]: `Char2` doesn't make a lot of sense now that `Char1` is not part of the public interface any more

File: std/string.d [hunk 3]
```diff
@@ -581,29 +581,90 @@ ptrdiff_t indexOf(Char)(const(Char)[] s, in dchar c, in size_t startIdx,
...
+
+        const(Char1)[] balance;
+        if (cs == CaseSensitive.yes)
+        {
+            balance = std.algorithm.find(s, sub);
+        }
+        else
+        {
+            balance = std.algorithm.find!
+                ((a, b) => std.uni.toLower(a) == std.uni.toLower(b))
```
Comment (JakobOvrum) [inline std/string.d]: Although tangential to this PR as it's from the old code, this isn't actually correct, as not all code points have 1:1 upper:lower mapping.

File: std/string.d [hunk 4]
```diff
@@ -581,29 +581,90 @@ ptrdiff_t indexOf(Char)(const(Char)[] s, in dchar c, in size_t startIdx,
...
-    $(D cs) indicates whether the comparisons are case sensitive.
+    Params:
+        s = string or ForwardRange of characters to search in correct UTF format
+        sub = substring to search for
+        cs = CaseSensitive.yes or CaseSensitive.no
+
+    Returns:
+        the index of the first occurrence of $(D sub) in $(D s). If $(D sub)
+        is not found, then $(D -1) is returned.
+        If the parameters are not valid UTF, the result will still
```
Comment (JakobOvrum) [inline std/string.d]: nitpick: s/parameters/arguments/

**Comment by JakobOvrum** [comment]
LGTM

File: std/string.d (hunk 3 above)
Comment (WalterBright) [inline std/string.d]: Yeah, added a `Bugs` notation.

File: std/string.d (hunk 2 above)
Comment (WalterBright) [inline std/string.d]: changed to `Char`

File: std/string.d (hunk 4 above)
Comment (WalterBright) [inline std/string.d]: done

**Comment by JakobOvrum** [comment]
Auto-merge toggled on

**Pull request merged**